django/template/defaulttags.py

```diff
@@ -1188,21 +1189,36 @@ def querystring(context, query_dict=None, **kwargs):
 
         {% querystring page=page_obj.next_page_number %}
 
-    A custom ``QueryDict`` can also be used::
+    A custom ``QueryDict`` can be used::
 
         {% querystring my_query_dict foo=3 %}
+
+    A custom ``dict`` can also be used::
+
+        {% querystring my_dict foo=3 %}
+
+    And a mix of multiple args and kwargs::
+
+        {% querystring my_dict my_query_dict foo=3 bar=None %}
     """
-    if query_dict is None:
-        query_dict = context.request.GET
-    query_dict = query_dict.copy()
-    for key, value in kwargs.items():
-        if value is None:
-            if key in query_dict:
-                del query_dict[key]
-        elif isinstance(value, Iterable) and not isinstance(value, str):
-            query_dict.setlist(key, value)
-        else:
-            query_dict[key] = value
+    try:
+        query_dict = context.request.GET.copy()
+    except AttributeError:
+        if not args:
+            raise
+        query_dict = QueryDict(mutable=True)
+    for d in args + (kwargs,):
+        for key, value in d.items():
+            if not isinstance(key, str):
+                raise TemplateSyntaxError(
+                    "querystring received non-string dict key: %s" % key
+                )
+            if value is None:
+                query_dict.pop(key, None)
+            elif isinstance(value, Iterable) and not isinstance(value, str):
+                query_dict.setlist(key, value)
+            else:
+                query_dict[key] = value
```

Comment:
I am not sure if we want more validation on values but here but here is another possible dictionary input and it's output for a test
```python
    @setup({"querystring_dict_value": "{% querystring my_dict %}"})
    def test_querystring_dict_value(self):
        context = {"my_dict": {"a": {"a": 0}}}
        output = self.engine.render_to_string("querystring_dict_value", context)
        self.assertEqual(output, "?a=a")
```
Is that what we want? Or do we want it to raise an error do you think? :thinking:

Reply:
That is interesting. In my opinion this specific case kind of makes sense, it's not really unexpected to me. The inner dict is treated as an iterable by the tag and `QueryDict.setlist()` is used.
What definitely is more problematic is passing arbitrary objects as dict values:

```python
@setup({"querystring_class_value": "{% querystring my_dict %}"})
def test_querystring_class_value(self):
    class MyClass:
        pass

    context = {"my_dict": {"a": MyClass()}}
    output = self.engine.render_to_string("querystring_class_value", context)
    print(output)
```
Which outputs `?a=%3Ctemplate_tests.syntax_tests.test_querystring.QueryStringTagTests.test_querystring_dict_value.%3Clocals%3E.MyClass+object+at+0x72a2e4f19990%3E`

On the other hand I'm not sure how we can limit this properly. Could we only allow string and numeric dict values?
And then what happens with Iterables, do we also check the types of their contents too?
Maybe it's "ok" to allow other values, given that in the cases that I can think of, the tag will either raise an exception, or the output will be obviously, noticeably wrong to the user.

Reply:
There is one use case for using, for example, int values as keys which I added to my former review comment (reproduced below):

> Also my dict `{i: i*2 for i in range(5)}` does not work either, it fails with `querystring received non-string dict key: 0`, I wonder if we should use `str()` of the keys, and only fail if `str()` fails?

So resurfacing that here: Is there a strong objection to applying `str()` to every key?

Reply:
On  a second look and thought, I agree we can tweak this after feature freeze following developer feedback. Will resolve for now,